File: go-version/leetcode/LC2799/count_complete_subarrays.go
package LC2799

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/4/24 下午11:12
 * @FilePath: count_complete_subarrays
 * @Description: 滑动窗口
 */

func countCompleteSubarrays(nums []int) int {
	distinct := differentNums(nums)
	ans, left := 0, 0
	mp := make(map[int]int)
	for i := range nums {
		mp[nums[i]]++
		for len(mp) >= distinct {
			mp[nums[left]]--
			if mp[nums[left]] == 0 {
				delete(mp, nums[left])
			}
			left++
		}
		ans += left
	}
	return ans
}

func differentNums(nums []int) int {
	mp := make(map[int]struct{})
	for _, num := range nums {
		mp[num] = struct{}{}
	}
	return len(mp)
}

File: python-version/LC2799/count_complete_subarrays.py
class Solution:
    def countCompleteSubarrays(self, nums: list[int]) -> int:
        left, ans = 0, 0
        distinct = self.__different_nums(nums)
        mp = dict()
        for i, _ in enumerate(nums):
            mp[nums[i]] = mp.get(nums[i], 0) + 1
            while len(mp) >= distinct:
                mp[nums[left]] = mp.get(nums[left]) - 1
                if mp[nums[left]] == 0:
                    del mp[nums[left]]
                left += 1
            ans += left
        return ans

    @staticmethod
    def __different_nums(nums: list[int]) -> int:
        st = set()
        for num in nums:
            st.add(num)
        return len(st)
